File: 16-05-2024_Divisibility_tree.java
class Solution {
    private int ans = 0; // declare globally
    
    private int dfs(int start, List<List<Integer>> adj, boolean[] visited) {
        int count = 0; // storing size of subtree
        int result = 0;
        visited[start] = true;
        for (int it : adj.get(start)) {
            if (!visited[it]) {
                result = dfs(it, adj, visited); // result is storing size of subtree
                if (result % 2 == 0) {
                    ans++;
                } else {
                    count += result;
                }
            }
        }
        return count + 1;
    }
    
    public int minimumEdgeRemove(int n, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }
        boolean[] visited = new boolean[n];
        
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj.get(u - 1).add(v - 1);
            adj.get(v - 1).add(u - 1);
        }
        
        dfs(0, adj, visited);
        return ans;
    }
}

File: 16-05-2024_Divisibility_tree.cpp
class Solution {
public:
    int ans=0; //declare globally
    
    int dfs(int start,vector<vector<int>>&adj,vector<int>&visited){
         int count=0; //storing size of subtree
         int result=0;
         visited[start]=1;
         for(auto it: adj[start]){
             if(!visited[it]){
                 result= dfs(it,adj,visited);  //result is storing size of subtree
                 if(result%2==0){
                     ans++;
                 }
                 else{
                     count+=result;
                 }
             }
         }
         return count+1;
    }


	int minimumEdgeRemove(int n, vector<vector<int>>edges){
	    vector<vector<int>>adj(n+1);
	    vector<int>visited(n,0);
	
	    for(auto edge: edges){
	        int u=edge[0];
	        int v=edge[1];
	        adj[u-1].push_back(v-1);
	        adj[v-1].push_back(u-1);
	    }
	   
	    dfs(0,adj,visited);
	    return ans;
	}
};
